### ISSUE TITLE
systemd-shutdown fails to stop mdraid on top of dm_integrity volumes

### ISSUE THREAD

**Issue by tomato42** [systemd/systemd#17103]
**systemd version the issue has been seen with**
> systemd 246.4

**Used distribution**
 > Archlinux

**Expected behaviour you didn't see**
 > clean shutdown

**Unexpected behaviour you saw**
 > systemd-shutdown didn't manage to stop and cleanly unmount lvm2 on mdraid on top of dm_integrity volumes

**Steps to reproduce the problem**
1. set up dm_integrity volumes:
   ```
   integritysetup format /dev/sdb1 --sector-size 4096 --tag-size 8 --integrity sha1
   integritysetup format /dev/sdc1 --sector-size 4096 --tag-size 8 --integrity sha1
   integritysetup format /dev/sdd1 --sector-size 4096 --tag-size 8 --integrity sha1
   integritysetup format /dev/sde1 --sector-size 4096 --tag-size 8 --integrity sha1 
   integritysetup format /dev/sdf1 --sector-size 4096 --tag-size 8 --integrity sha1
   integritysetup format /dev/sdg1 --sector-size 4096 --tag-size 8 --integrity sha1
   ```
2. Start dm_integrity volumes:
   ```
   integritysetup open /dev/sdb1 int-sdb
   integritysetup open /dev/sdc1 int-sdc
   integritysetup open /dev/sdd1 int-sdd
   integritysetup open /dev/sde1 int-sde
   integritysetup open /dev/sdf1 int-sdf
   integritysetup open /dev/sdg1 int-sdg
   ```
3. Create and mount the mdraid:
   ```
   mdadm -C /dev/md0 -l 6 -n 6 /dev/mapper/int-sd?
   mdadm --wait /dev/md0
   ```
4. Create the lvm and an LV:
   ```
   pvcreate /dev/md0
   vgcreate raid-test /dev/md0
   lvcreate -l 2G -n exports raid-test
   ```
5. mkfs and mount:
   ```
   mkfs.ext4 /dev/raid-test/exports
   mkdir /exports
   mount /dev/raid-test/exports /exports
   ```
6. reboot

The last lines from shutdown-log.txt are:
```
[  248.350992] systemd[430]: Closed p11-kit server.
[  248.351407] systemd[430]: shutdown.target changed dead -> active
[  248.351810] systemd[430]: shutdown.target: Job 20 shutdown.target/start finished, result=done
[  248.352255] systemd[430]: Reached target Shutdown.
[  248.352685] systemd[430]: systemd-exit.service: Changed dead -> start
[  248.353122] systemd[430]: systemd-exit.service: Succeeded.
[  248.353537] systemd[430]: systemd-exit.service: Service will not restart (restart setting)
[  248.353933] systemd[430]: systemd-exit.service: Changed start -> dead
[  248.354349] systemd[430]: systemd-exit.service: Job 19 systemd-exit.service/start finished, result=done
[  248.355132] systemd[430]: Finished Exit the Session.
[  248.355528] systemd[430]: Exiting immediately: unit systemd-exit.service succeeded
[  248.355958] systemd[430]: exit.target changed dead -> active
[  248.356338] systemd[430]: exit.target: Job 18 exit.target/start finished, result=done
[  248.356722] systemd[430]: Reached target Exit the Session.
[  248.357093] systemd[430]: Exit.
[  248.358127] systemd[430]: Bus bus-system: changing state RUNNING \xe2\x86\x92 CLOSED
[  248.360910] audit: type=1131 audit(1600476784.420:41): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=user@0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[  248.366704] systemd-user-runtime-dir[505]: Will remove /run/user/0
[  248.369729] audit: type=1131 audit(1600476784.430:42): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=user-runtime-dir@0 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[  248.378972] audit: type=1131 audit(1600476784.436:43): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-user-sessions comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[  248.387973] audit: type=1131 audit(1600476784.446:44): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=dbus comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[  248.484278] audit: type=1131 audit(1600476784.543:45): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=dhcpcd comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[  248.503440] audit: type=1131 audit(1600476784.563:46): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-sysctl comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[  248.509737] systemd-update-utmp[509]: Bus n/a: changing state UNSET \xe2\x86\x92 OPENING
[  248.510160] systemd-update-utmp[509]: Bus n/a: changing state OPENING \xe2\x86\x92 AUTHENTICATING
[  248.511498] systemd-update-utmp[509]: Not writing wtmp: No such file or directory
[  248.511869] systemd-update-utmp[509]: Bus n/a: changing state AUTHENTICATING \xe2\x86\x92 RUNNING
[  248.512236] systemd-update-utmp[509]: Bus n/a: changing state RUNNING \xe2\x86\x92 CLOSED
[  249.577282] systemd-journald[323]: Failed to send stream file descriptor to service manager: Connection refused
[  249.578144] systemd-journald[323]: Failed to send stream file descriptor to service manager: Transport endpoint is not connected
[  249.579179] systemd-journald[323]: Failed to send stream file descriptor to service manager: Transport endpoint is not connected
[  249.580109] systemd-journald[323]: Failed to send stream file descriptor to service manager: Transport endpoint is not connected
[  249.599994] systemd-shutdown[1]: Found cgroup2 on /sys/fs/cgroup/unified, unified hierarchy for systemd controller
[  249.601195] systemd-journald[323]: Failed to send stream file descriptor to service manager: Transport endpoint is not connected
[  249.616145] systemd-journald[323]: Failed to send stream file descriptor to service manager: Transport endpoint is not connected
[  249.617304] systemd-journald[323]: Failed to send stream file descriptor to service manager: Transport endpoint is not connected
[  252.351804] systemd-shutdown[1]: Successfully forked off '(sd-sync)' as PID 514.
[  252.352264] systemd-shutdown[1]: Syncing filesystems and block devices.
[  252.833317] systemd-shutdown[1]: binfmt_misc is not mounted, not detaching entries.
[  252.833769] systemd-shutdown[1]: Sending SIGTERM to remaining processes...
[  252.835496] systemd-journald[323]: Received SIGTERM from PID 1 (systemd-shutdow).
[  252.836065] systemd-journald[323]: systemd-journald stopped as PID 323 for the system.
[  252.836607] systemd-journald[323]: Assertion 's->event->state != SD_EVENT_FINISHED' failed at src/libsystemd/sd-event/sd-event.c:2386, function sd_event_source_set_time(). Ignoring.
[  252.837494] systemd-journald[323]: Failed to set time for scheduling ftruncate: Stale file handle
[  253.314494] kauditd_printk_skb: 16 callbacks suppressed
[  253.314495] audit: type=1335 audit(1600476789.373:63): pid=323 uid=0 auid=4294967295 tty=(none) ses=4294967295 comm="systemd-journal" exe="/usr/lib/systemd/systemd-journald" nl-mcgrp=1 op=disconnect res=1
[  253.316757] systemd-shutdown[1]: Sending SIGKILL to remaining processes...
[  253.318651] systemd-shutdown[1]: Unmounting file systems.
[  253.319446] systemd-shutdown[1]: Successfully forked off '(sd-remount)' as PID 515.
[  253.320067] [515]: Remounting '/' read-only in with options '(null)'.
[  254.097118] EXT4-fs (sda1): re-mounted. Opts: (null)
[  254.098785] systemd-shutdown[1]: All filesystems unmounted.
[  254.099317] systemd-shutdown[1]: Deactivating swaps.
[  254.099840] systemd-shutdown[1]: All swaps deactivated.
[  254.100319] systemd-shutdown[1]: Detaching loop devices.
[  254.100788] systemd-shutdown[1]: sd-device-enumerator: Scan all dirs
[  254.101262] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/bus
[  254.101726] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/class
[  254.102194] systemd-shutdown[1]: All loop devices detached.
[  254.102645] systemd-shutdown[1]: Detaching DM devices.
[  254.103055] systemd-shutdown[1]: sd-device-enumerator: Scan all dirs
[  254.103468] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/bus
[  254.103865] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/class
[  254.104717] systemd-shutdown[1]: Detaching DM /dev/dm-6 (254:6).
[  254.124710] systemd-shutdown[1]: Detaching DM /dev/dm-5 (254:5).
[  254.125091] systemd-shutdown[1]: Could not detach DM /dev/dm-5: Device or resource busy
[  254.125442] systemd-shutdown[1]: Detaching DM /dev/dm-4 (254:4).
[  254.125787] systemd-shutdown[1]: Could not detach DM /dev/dm-4: Device or resource busy
[  254.126123] systemd-shutdown[1]: Detaching DM /dev/dm-3 (254:3).
[  254.126455] systemd-shutdown[1]: Could not detach DM /dev/dm-3: Device or resource busy
[  254.126791] systemd-shutdown[1]: Detaching DM /dev/dm-2 (254:2).
[  254.127121] systemd-shutdown[1]: Could not detach DM /dev/dm-2: Device or resource busy
[  254.127446] systemd-shutdown[1]: Detaching DM /dev/dm-1 (254:1).
[  254.127777] systemd-shutdown[1]: Could not detach DM /dev/dm-1: Device or resource busy
[  254.128105] systemd-shutdown[1]: Detaching DM /dev/dm-0 (254:0).
[  254.128434] systemd-shutdown[1]: Could not detach DM /dev/dm-0: Device or resource busy
[  254.128763] systemd-shutdown[1]: Not all DM devices detached, 6 left.
[  254.129082] systemd-shutdown[1]: Couldn't finalize remaining  DM devices, trying again.
[  254.129481] systemd-shutdown[1]: Detaching DM devices.
[  254.129789] systemd-shutdown[1]: sd-device-enumerator: Scan all dirs
[  254.130099] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/bus
[  254.130398] systemd-shutdown[1]: sd-device-enumerator: Scanning /sys/class
[  254.131045] systemd-shutdown[1]: Detaching DM /dev/dm-5 (254:5).
[  254.131320] systemd-shutdown[1]: Could not detach DM /dev/dm-5: Device or resource busy
[  254.131593] systemd-shutdown[1]: Detaching DM /dev/dm-4 (254:4).
[  254.131865] systemd-shutdown[1]: Could not detach DM /dev/dm-4: Device or resource busy
[  254.132134] systemd-shutdown[1]: Detaching DM /dev/dm-3 (254:3).
[  254.132402] systemd-shutdown[1]: Could not detach DM /dev/dm-3: Device or resource busy
[  254.132672] systemd-shutdown[1]: Detaching DM /dev/dm-2 (254:2).
[  254.132945] systemd-shutdown[1]: Could not detach DM /dev/dm-2: Device or resource busy
[  254.133222] systemd-shutdown[1]: Detaching DM /dev/dm-1 (254:1).
[  254.133511] systemd-shutdown[1]: Could not detach DM /dev/dm-1: Device or resource busy
[  254.133795] systemd-shutdown[1]: Detaching DM /dev/dm-0 (254:0).
[  254.134087] systemd-shutdown[1]: Could not detach DM /dev/dm-0: Device or resource busy
[  254.134385] systemd-shutdown[1]: Not all DM devices detached, 6 left.
[  254.134683] systemd-shutdown[1]: Cannot finalize remaining DM devices, continuing.
[  254.135109] systemd-shutdown[1]: Successfully forked off '(sd-executor)' as PID 516.
[  254.153500] [516]: Successfully forked off '(direxec)' as PID 517.
[  254.153943] [516]: Successfully forked off '(direxec)' as PID 518.
[  254.233214] [516]: /usr/lib/systemd/system-shutdown/mdadm.shutdown succeeded.
[  254.323079] EXT4-fs (sda1): re-mounted. Opts: (null)
```

I've added few more commands to the debug.sh and `/proc/mdstat` reports the array as active:
```
Personalities : [raid6] [raid5] [raid4] 
               md0 : active raid6 dm-5[12] dm-4[11] dm-3[10] dm-2[9] dm-1[8] dm-0[7]
                     83415040 blocks super 1.2 level 6, 512k chunk, algorithm 2 [6/6] [UUUUUU]
                     
               unused devices: <none>
```
But I see no reason why systemd-shutdown wouldn't be able to stop it, there are no holders of that block device:
```
# ls /sys/block/md0/holders/
#
```

Side note: `mdadm.shutdown` just calls `/sbin/mdadm --wait-clean --scan`

**Comment by tomato42** [comment]
full [shutdown-log.txt](https://github.com/systemd/systemd/files/5248914/shutdown-log.txt) file

**Comment by arvidjaar** [comment]
> But I see no reason why `systemd-shutdown` wouldn't be able to stop it

It probably could stop it but currently it does not handle MD arrays at all.

`[  254.233214] [516]: /usr/lib/systemd/system-shutdown/mdadm.shutdown succeeded.`

It is too late. Also `mdadm.shutdown` is not intended to stop layered block devices - it is intended to flush firmware RAID state so  array becomes clean on reboot. Firmware RAIDs are not placed on device-mapper nodes.

The subject is misleading. It has nothing to do with integrity. Any layered storage stack with MD array on top would have the same issue (even it would have been simple loop device).

**Comment by tomato42** [comment]
> > But I see no reason why `systemd-shutdown` wouldn't be able to stop it
> 
> It probably could stop it but currently it does not handle MD arrays at all.

why? that's a design decision or a lacking feature?
 
> `[ 254.233214] [516]: /usr/lib/systemd/system-shutdown/mdadm.shutdown succeeded.`
> 
> It is too late. 

how's it too late? `/dev/dm-6` is the LV in LVM and it got deactivated just before that

**Comment by poettering** [comment]
the loop in systemd-shutdown that detaches mounts/loop devices/DM devices is a safety net. That's all. It's not supposed to be able to cover everything, but just a few simple things.

Storage subsystems should detach their own stuff:

1. Either with a system service's ExecStop= for storage volumes that are not needed for the init system itself
2. Or via the initrd that systemd jumps back into during late shutdown, to cover for storage volumes that the init system is backed by itself (i.e. the root volume).

Either way, systemd is not the responsible for detaching all kinds of storage, that's really a job for the storage packages. The fact that we implement some simple safety net, is just to add robustness, not that we think it's our responsibility to implement storage dissassembly.

Long story short: please talk to the maintainers of the storage subsystems you are making use of (i.e. mdadm, integritysetup) for help about this.

(That all said: if there's a simple ioctl we can call to detach MD devices in our safety net, similar to how we already do it for DM devices, then we can certainly make use of that, happy to review/merge a patch for that, if anyone cares.)

Either, let's close this here, there's nothing to fix here in systemd.
